zerver/lib/send_email.py

```diff
@@ -441,8 +442,7 @@ def deliver_scheduled_emails(email: ScheduledEmail) -> None:
         data["to_user_ids"] = user_ids
     if email.address is not None:
         data["to_emails"] = [email.address]
-    handle_send_email_format_changes(data)
-    send_email(**data)
+    queue_json_publish("email_senders", data)
```

Comment:
Do we want to do a `try/except` here and only enqueue the data if we get a failure?

Reply:
This would open a new SMTP connection while if we push it directly to the queue it will use a previously opened one (although it might be refreshed) but this could avoid to use the worker when it is not necessary to have a retry mechanism because it is correctly sent at first try. I'm not sure wich way would be the most efficient ?

Reply:
I'd rather only have one process in charge of talking SMTP.  As-is, SMTP failures can show up in either `events_deliver_scheduled_emails` or in `events_email_sender`.  When we move to async workers, there isn't much reason to make `deliver_scheduled_emails` async -- unless it's doing network IO, but that may complicate things.

I think it suits the UNIX philosophy to have this use the `email_senders` queue as its API for sending.

Reply:
I think that's a good high-level theory, but I think that would have the side effect of dumping digest emails into the main queue, which we really don't want with current throughput levels.

(They're sent with `send_future_email` for that reason, I think).

Reply:
Does it mean we should abandon these changes ? I am not sure if you both agreed on giving up or if you're trying to find a compromise in terms of performances ?

Reply:
Well, I think a version of this that put things in `email_senders` on error could be useful; I'm curious what Alex thinks.  

It might mean we just want to put this on hold until we make the `email_senders` queue much higher-throughput, by making it an asynchronous queue worker.

Reply:
I don't think there's much utility to sending only errors through the main queue -- that just seems ripe for confusion.

Digest emails going through `send_future_email` (and not through the main sending path) seems to mostly be an accident of history (772adc1969649df17bc39eea3813704ef65f1a02) and is ... not well-documented.  I agree there's a little value to thinking about QOS rather than dumping everything into a single queue, but improving the throughput on the queue first makes more sense than doing any work on that.

So yeah, happy to punt on this until we get the sending async.  I don't think we should make errors temporarily go through the main queue is worth anything -- it mostly just muddles things.

Reply:
Oh right. Is there anything I can do about getting the sending async ? Maybe it's something I could work on as we put this commit aside for now ?